src/fs/bootstrap_record.cc

```diff
@@ -0,0 +1,187 @@
+            return make_exception_future<bootstrap_record>(
+                    invalid_bootstrap_record("Sector size should be a power of 2 and greater than 512"));
+        }
+        if (!is_cluster_size_valid(boot_record_disk.cluster_size, boot_record_disk.sector_size)) {
+            return make_exception_future<bootstrap_record>(
+                    invalid_bootstrap_record("Cluster size should be a power of 2 and be divisible by sector size"));
+        }
+
+        const std::vector<shard_info> shards_info(boot_record_disk.shards_info,
+                boot_record_disk.shards_info + boot_record_disk.shards_nb);
+
+        bootstrap_record boot_record_mem(boot_record_disk.version,
+                boot_record_disk.sector_size,
+                boot_record_disk.cluster_size,
+                boot_record_disk.root_directory,
+                boot_record_disk.shards_nb,
+                std::move(shards_info));
+
+        return make_ready_future<bootstrap_record>(boot_record_mem);
+    });
+}
+
+future<> bootstrap_record::write_to_disk(block_device& device) const {
+    // initial check
+    if (shards_info.size() > max_shards_nb) {
+        return make_exception_future<>(invalid_bootstrap_record("Too many shards to fit into the bootstrap record"));
+    } else if (shards_nb != shards_info.size()) {
+        return make_exception_future<>(
+                invalid_bootstrap_record("Shards number should match number of metadata pointers"));
+    }
+    if (!is_sector_size_valid(sector_size)) {
+        return make_exception_future<>(invalid_bootstrap_record("Sector size should be a power of 2 and greater than 512"));
+    }
+    if (!is_cluster_size_valid(cluster_size, sector_size)) {
+        return make_exception_future<>(invalid_bootstrap_record("Cluster size should be a power of 2 and be divisible by sector size"));
+    }
+
+    bootstrap_record_disk boot_record_disk;
+    std::memset(&boot_record_disk, 0, sizeof(boot_record_disk));
+
+    // prepare bootstrap_record_disk records
+    boot_record_disk.magic = bootstrap_record::magic_number;
+    boot_record_disk.version = version;
+    boot_record_disk.sector_size = sector_size;
+    boot_record_disk.cluster_size = cluster_size;
+    boot_record_disk.root_directory = root_directory;
+    boot_record_disk.shards_nb = shards_nb;
+    std::copy(shards_info.begin(), shards_info.end(), boot_record_disk.shards_info);
+
+    auto boot_record_buff = temporary_buffer<char>::aligned(alignment, aligned_boot_record_size);
+    std::memcpy(boot_record_buff.get_write(), &boot_record_disk, sizeof(boot_record_disk));
+    std::memset(boot_record_buff.get_write() + sizeof(boot_record_disk), 0,
+            aligned_boot_record_size - sizeof(boot_record_disk));
+
+    boot_record_disk.crc = crc32(boot_record_buff.get(), crc_offset);
+    std::memcpy(boot_record_buff.get_write() + crc_offset, &boot_record_disk.crc, sizeof(boot_record_disk.crc));
```

Comment:
This can be done before allocating boot_record_buff by replacing `boot_record_buff.get()` with `&boot_record_disk`.